e2e/LongHaul/amqpPooling/IotHub.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Mash.Logging;
+using Microsoft.Azure.Devices.Client;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Devices.LongHaul.AmqpPooling
+{
+    internal class IotHub : IAsyncDisposable
+    {
+        private readonly string _hubConnectionString;
+        private readonly Logger _logger;
+        private readonly IotHubClientTransportSettings _transportSettings;
+        private readonly IotHubClientOptions _clientOptions;
+        private static IList<Device> s_devices;
+        private static IDictionary<string, DeviceOperations> s_deviceOperations;
+
+        private static readonly TimeSpan s_messageLoopSleepTime = TimeSpan.FromSeconds(3);
+        private readonly ConcurrentQueue<SystemHealthTelemetry> _messagesToSend = new();
+
+        private SemaphoreSlim _lifetimeControl = new(1, 1);
+
+        public IotHub(Logger logger, Parameters parameters, IList<Device> devices)
+        {
+            _logger = logger;
+            _hubConnectionString = parameters.IotHubConnectionString;
+            _transportSettings = parameters.GetTransportSettingsWithPooling();
+            _clientOptions = new IotHubClientOptions(_transportSettings);
+
+            s_devices = devices;
+            s_deviceOperations = new Dictionary<string, DeviceOperations>();
+        }
+
+        public Dictionary<string, string> TelemetryUserProperties { get; } = new();
+
+        public async Task InitializeAsync()
+        {
+            await _lifetimeControl.WaitAsync().ConfigureAwait(false);
+
+            var helper = new IotHubConnectionStringHelper(_hubConnectionString);
+
+            try
+            {
+                _logger.Trace(
+                    $"Creating {s_devices.Count} device clients with transport settings [{_transportSettings.ToString()}].",
+                    TraceSeverity.Information);
+
+                foreach (var device in s_devices)
+                {
+                    string deviceConnectionString = $"HostName={helper.HostName};DeviceId={device.Id};SharedAccessKey={device.Authentication.SymmetricKey.PrimaryKey}";
+
+                    var deviceClient = new IotHubDeviceClient(deviceConnectionString, _clientOptions);
+
+                    // Set the device client connection status change handler
+                    var amqpConnectionStatusChange = new AmqpConnectionStatusChange(_logger.Clone());
+                    deviceClient.ConnectionStatusChangeCallback = amqpConnectionStatusChange.ConnectionStatusChangeHandler;
+
+                    await deviceClient.OpenAsync().ConfigureAwait(false);
```

Comment:
Devices will need connection status callbacks to try to reconnect. Might as well put this call to open as well as that callback in the DeviceOperations class.

This code could simply create the devices and then kick off the operations which themselves handle all the device-specific concerns, including looping for telemetry.